grpc/shared/rollupdata_handling.go

```diff
@@ -0,0 +1,261 @@
+			//
+			// the fees are spent from the "bridge account" which is not actually a real account, but is instead some
+			// address defined by consensus, so the gas cost is not actually deducted from any account.
+			Gas:                    64000,
+			To:                     &bac.Erc20Asset.ContractAddress,
+			Data:                   calldata,
+			SourceTransactionId:    *deposit.SourceTransactionId,
+			SourceTransactionIndex: deposit.SourceActionIndex,
+		}
+
+		tx := types.NewTx(&txdata)
+		return tx, nil
+	}
+
+	txdata := types.DepositTx{
+		From:                   bac.SenderAddress,
+		To:                     &recipient,
+		Value:                  amount,
+		Gas:                    0,
+		SourceTransactionId:    *deposit.SourceTransactionId,
+		SourceTransactionIndex: deposit.SourceActionIndex,
+	}
+	return types.NewTx(&txdata), nil
+}
+
+func validateAndUnmarshallSequenceAction(tx *sequencerblockv1.RollupData) (*types.Transaction, error) {
+	ethTx := new(types.Transaction)
+	err := ethTx.UnmarshalBinary(tx.GetSequencedData())
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal sequenced data into transaction: %w. tx hash: %s", err, sha256.Sum256(tx.GetSequencedData()))
+	}
+
+	if ethTx.Type() == types.DepositTxType {
+		return nil, fmt.Errorf("deposit tx not allowed in sequenced data. tx hash: %s", sha256.Sum256(tx.GetSequencedData()))
+	}
+
+	if ethTx.Type() == types.BlobTxType {
+		return nil, fmt.Errorf("blob tx not allowed in sequenced data. tx hash: %s", sha256.Sum256(tx.GetSequencedData()))
+	}
+
+	return ethTx, nil
+}
+
+func unmarshallAllocationTxs(allocation *auctionv1alpha1.Allocation, prevBlockHash []byte, auctioneerBech32Address string, addressPrefix string) (types.Transactions, error) {
+	unbundlingStart := time.Now()
+	defer allocationUnbundlingTimer.UpdateSince(unbundlingStart)
+
+	processedTxs := types.Transactions{}
+	bid := &auctionv1alpha1.Bid{}
+
+	unprocessedBid := allocation.GetBid()
+
+	err := anypb.UnmarshalTo(unprocessedBid, bid, proto2.UnmarshalOptions{
+		Merge:        false,
+		AllowPartial: false,
+	})
+	if err != nil {
+		return nil, WrapError(err, "failed to unmarshal bid")
+	}
+
+	log.Debug("Found a potential allocation in the rollup data. Checking if it is valid.", "prevBlockHash", common.BytesToHash(prevBlockHash).String(), "auctioneerBech32Address", auctioneerBech32Address)
+
+	if !bytes.Equal(bid.GetRollupParentBlockHash(), prevBlockHash) {
+		allocationsWithInvalidPrevBlockHash.Inc(1)
+		return nil, errors.New("prev block hash in allocation does not match the previous block hash")
+	}
+
+	publicKey := ed25519.PublicKey(allocation.GetPublicKey())
+	bech32Address, err := EncodeFromPublicKey(addressPrefix, publicKey)
+	if err != nil {
+		return nil, WrapError(err, fmt.Sprintf("failed to encode public key to bech32m address: %s", publicKey))
+	}
+
+	if auctioneerBech32Address != bech32Address {
```

Comment:
why put public key bytes into the `allocation` instead of a bech32 string directly?